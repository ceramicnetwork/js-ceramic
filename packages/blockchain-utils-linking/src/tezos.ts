import { AccountId } from 'caip'
import { AuthProvider } from './auth-provider.js'
import { asOldCaipString, getConsentMessage, LinkProof } from './util.js'
import { hash } from '@stablelib/sha256'
import * as uint8arrays from 'uint8arrays'
import type { Signer } from '@taquito/taquito'

export const TEZOS_NAMESPACE = 'tezos'
export const TEZOS_CHAIN_REF = 'NetXdQprcVkpaWU' // Tezos mainnet

export interface TezosProvider {
  // the below is requried from TezosToolkit
  signer: Signer
}

/**
 * Encodes a message to its Micheline string representation
 *
 * @param {string} text - the message to Encode
 * @returns {string} the message encoded to the micheline format
 *
 * References:
 * - https://github.com/Cryptonomic/ConseilJS/blob/fb718632d6ce47718dad5aa77c67fc514afaa0b9/src/chain/tezos/lexer/Micheline.ts#L62L71
 * - https://tezos.gitlab.io/shell/micheline.html
 */
function encodeMessage(text: string): string {
  const michelinePrefix = '05'
  const stringPrefix = '01'
  const len = ('0000000' + text.length.toString(16)).slice(-8)

  text = uint8arrays.toString(uint8arrays.fromString(text, 'utf-8'), 'hex')
  return michelinePrefix + stringPrefix + len + text
}

/**
 * signs a message with the active address and returns the signature
 *
 * @param {TezosProvider} provider - the provider to use for signing
 * @param {string} message - the message to sign
 * @returns {Promise<string>} the signature prefixed with the signing method
 */
async function sign(provider: TezosProvider, message: string): Promise<string> {
  // sign the message with the active address and get the signature with the type prefixed
  message = encodeMessage(message)
  const { prefixSig: signature } = await provider.signer.sign(message)
  return signature
}

/**
 * get the active address from the {TezosProvider}
 *
 * @param {TezosProvider} provider - the provider to use for getting the active account
 * @returns {Promise<string>} - a promise that resolves to the active account's address
 */
async function getActiveAddress(provider: TezosProvider): Promise<string> {
  return provider.signer.publicKeyHash()
}

/**
 * This is an implementation of {AuthProvider} for the Tezos network.
 * It uses the Tezos RPC client to get the active account's address and sign the
 * message with the active account's address.
 *
 * @param {TezosProvider} provider - the provider to use signing the link proof and getting the active account's address
 */
export class TezosAuthProvider implements AuthProvider {
  readonly isAuthProvider = true

  constructor(private readonly provider: TezosProvider) {
    if (provider.signer === undefined) {
      throw new Error('a `Signer` is required to use the `TezosAuthProvider`')
    }

    console.warn(
      'WARN: TezosAuthProvider is not fully supported. You may encounter issues using this.'
    )
  }

  /**
   * @inheritdoc
   */
  async authenticate(message: string): Promise<string> {
    const signature = await sign(this.provider, message)
    const digest = hash(uint8arrays.fromString(signature))
    return `0x${uint8arrays.toString(digest, 'base16')}`
  }

  /**
   * @inheritdoc
   */
  async createLink(did: string): Promise<LinkProof> {
    // get consent message
    const { message, timestamp } = getConsentMessage(did)
    // sign the message with the active address
    const signature = await sign(this.provider, message)
    const address = await getActiveAddress(this.provider)
    // generate account ID
    const caipAccount = new AccountId({
      address,
      chainId: `${TEZOS_NAMESPACE}:${TEZOS_CHAIN_REF}`,
    })
    // create link proof
    const proof: LinkProof = {
      version: 2,
      message,
      signature,
      account: asOldCaipString(caipAccount),
      timestamp,
    }
    return proof
  }

  /**
   * @inheritdoc
   */
  async accountId(): Promise<AccountId> {
    const address = await getActiveAddress(this.provider)
    return new AccountId({
      address,
      chainId: `${TEZOS_NAMESPACE}:${TEZOS_CHAIN_REF}`,
    })
  }

  /**
   * This is unsupported
   *
   * @throws {Error} - always throws an error
   */
  withAddress(_address: string): AuthProvider {
    throw new Error('TezosAuthProvider does not support withAddress')
  }
}
