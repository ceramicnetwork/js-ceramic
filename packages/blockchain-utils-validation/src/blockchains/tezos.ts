import { ChainId } from 'caip'
import { verifySignature } from '@taquito/utils'
import { BlockchainHandler } from '../blockchain-handler.js'
import { LinkProof, tezos } from '@ceramicnetwork/blockchain-utils-linking'
import fetch from 'cross-fetch'
import * as uint8arrays from 'uint8arrays'
import { normalizeAccountId } from '@ceramicnetwork/common'

export const ADDRESS_NOT_FOUND_ERROR = new Error(`Address not found on the Tezos blockchain`)
export const PUBLIC_KEY_NOT_PUBLISHED_ERROR = new Error(
  `Public key not published to the Tezos blockchain`
)

// eslint-disable-next-line @typescript-eslint/ban-ts-comment
// @ts-ignore TS can't resolve these, but they are there and tests pass
const { TEZOS_NAMESPACE, TEZOS_CHAIN_REF } = tezos

/**
 * Encodes a message to its Micheline string representation
 *
 * @param {string} text - the message to Encode
 * @returns {string} the message encoded to the micheline format
 *
 * References:
 * - https://github.com/Cryptonomic/ConseilJS/blob/fb718632d6ce47718dad5aa77c67fc514afaa0b9/src/chain/tezos/lexer/Micheline.ts#L62L71
 * - https://tezos.gitlab.io/shell/micheline.html
 */
function encodeMessage(text: string): string {
  const michelinePrefix = '05'
  const stringPrefix = '01'
  const len = ('0000000' + text.length.toString(16)).slice(-8)

  text = uint8arrays.toString(uint8arrays.fromString(text, 'utf-8'), 'hex')
  return michelinePrefix + stringPrefix + len + text
}

/**
 * attempts to resolve the public key associated with the wallet address
 *
 * @param {string} address - a Tezos wallet address to lookup
 * @returns {Promise<string>} - a promise for the public key of the address
 * @throws {Error} - if the address is not found on the blockchain
 * @throws {Error} - if the public key is not published to the blockchain
 */
async function findPublicKey(address: string): Promise<string> {
  // request the public key from the Tezos blockchain
  const response = await fetch(`https://api.tzstats.com/explorer/account/${address}`).catch(() => {
    throw ADDRESS_NOT_FOUND_ERROR
  })

  const json = await response.json()
  const result = json?.pubkey
  if (result) {
    return result
  } else {
    throw PUBLIC_KEY_NOT_PUBLISHED_ERROR
  }
}

// can only validate proofs with wallet addresses (public key hashes) that have:
// - at least one transaction on the blockchain
// - their public key published to the blockchain
export async function validateLink(proof: LinkProof): Promise<LinkProof | null> {
  const account = normalizeAccountId(proof.account)
  const chainId = new ChainId(account.chainId)

  // only support Tezos mainnet for now
  if (chainId.reference !== TEZOS_CHAIN_REF) {
    return null
  }
  const msg = encodeMessage(proof.message)

  try {
    const pk = await findPublicKey(account.address)
    const is_sig_valid: boolean = await verifySignature(msg, pk, proof.signature)

    return is_sig_valid ? proof : null
  } catch (ignored) {
    return null
  }
}

export const handler: BlockchainHandler = {
  namespace: TEZOS_NAMESPACE,
  validateLink,
}
