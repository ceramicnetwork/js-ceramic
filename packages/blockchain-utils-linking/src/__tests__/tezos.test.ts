import { TezosAuthProvider, TezosProvider } from '../tezos.js'
import { InMemorySigner } from '@taquito/signer'

const did = 'did:3:bafysdfwefwe'
const privateKey = 'p2sk2obfVMEuPUnadAConLWk7Tf4Dt3n4svSgJwrgpamRqJXvaYcg1'
const chainRef = 'NetXdQprcVkpaWU' // Tezos mainnet
let provider: TezosProvider

// cache Date.now() to restore it after all tests
const dateNow = Date.now

beforeAll(async () => {
  // Mock Date.now() to return a constant value
  Date.now = () => 666000

  const signer = await InMemorySigner.fromSecretKey(privateKey)
  provider = {
    signer,
  }
})

afterAll(() => {
  // restore Date.now()
  Date.now = dateNow
})

describe('Blockchain: Tezos', () => {
  describe('createLink', () => {
    test(`create proof for ${chainRef}`, async () => {
      const authProvider = new TezosAuthProvider(provider)
      const proof = await authProvider.createLink(did)
      expect(proof).toMatchSnapshot()
    })
  })

  describe('authenticate', () => {
    test(`create proof for ${chainRef}`, async () => {
      const authProvider = new TezosAuthProvider(provider)
      const result = await authProvider.authenticate('msg')
      expect(result.length).toBe(2 + 2 * 32) // 2 for "0x" prefix, 2 * 32 for 32 bytes of entropy as a hex string
      expect(result).toMatchSnapshot()
    })
  })

  describe('authenticate', () => {
    test(`entropy replication for ${chainRef}`, async () => {
      const authProvider = new TezosAuthProvider(provider)
      const msg = 'hello'
      const result1 = await authProvider.authenticate(msg)
      const result2 = await authProvider.authenticate(msg)
      expect(result1).toMatchSnapshot()
      expect(result2).toMatchSnapshot()
    })
  })
})
