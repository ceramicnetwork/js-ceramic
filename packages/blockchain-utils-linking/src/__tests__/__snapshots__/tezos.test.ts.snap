// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Blockchain: Tezos authenticate create proof for NetXdQprcVkpaWU 1`] = `"0xa7be667195657a51d34efc8ce0fde68560d9110b243e737e695dc1325a9e8dbf"`;

exports[`Blockchain: Tezos authenticate entropy replication for NetXdQprcVkpaWU 1`] = `"0x56db1b9d98064056438a5a9245c9be1e5116d63493959bb8387198d7dc5111a0"`;

exports[`Blockchain: Tezos authenticate entropy replication for NetXdQprcVkpaWU 2`] = `"0x56db1b9d98064056438a5a9245c9be1e5116d63493959bb8387198d7dc5111a0"`;

exports[`Blockchain: Tezos createLink create proof for NetXdQprcVkpaWU 1`] = `
Object {
  "account": "tz3Lfm6CyfSTZ7EgMckptZZGiPxzs9GK59At@tezos:NetXdQprcVkpaWU",
  "message": "Link this account to your identity

did:3:bafysdfwefwe 
Timestamp: 666",
  "signature": "p2sigNYcqkNERyCnBb7PKwmuZqeXBZx4m2c8YKcR282iWN9iafSdcTrJ4VxQdANg1oo7nGteZ7uGEDiy1ZpAqCCb71xdy9MptM",
  "timestamp": 666,
  "version": 2,
}
`;
